src/main/java/io/appium/java_client/remote/AppiumCommandExecutor.java

```diff
@@ -111,8 +130,77 @@ private void setCommandCodec(CommandCodec<HttpRequest> newCodec) {
+
+                //the number of bytes before the stream should switch to buffering to a file
+                int threshold = (int) Math.min(Runtime.getRuntime().freeMemory() / 10, Integer.MAX_VALUE);
+                FileBackedOutputStream os = new FileBackedOutputStream(threshold);
+                try {
+
+                    CountingOutputStream counter = new CountingOutputStream(os);
+                    Writer writer = new OutputStreamWriter(counter, UTF_8);
+                    NewAppiumSessionPayload payload = NewAppiumSessionPayload.create(desired);
+                    payload.writeTo(writer);
+
+                    try (InputStream rawIn = os.asByteSource().openBufferedStream();
+                         BufferedInputStream contentStream = new BufferedInputStream(rawIn)) {
+
+                        Method createSessionMethod = this.getClass().getSuperclass()
+                                .getDeclaredMethod("createSession", HttpClient.class, InputStream.class, long.class);
+                        createSessionMethod.setAccessible(true);
+
+                        Optional<Result> result = (Optional<Result>) createSessionMethod
+                                .invoke(this, client, contentStream, counter.getCount());
+
+                        return result.map(result1 -> {
+                            Result toReturn = result.get();
+                            System.out.print(format("Detected dialect: %s", toReturn.getDialect()));
+                            return toReturn;
+                        }).orElseThrow(() -> new SessionNotCreatedException(
+                                format("Unable to create new remote session. desired capabilities = %s", desired)));
+                    } catch (NoSuchMethodException | IllegalAccessException | InvocationTargetException e) {
+                        throw new WebDriverException(format("It is impossible to create a new session "
+                                        + "because 'createSession' which takes %s, %s and %s was not found "
+                                        + "or it is not accessible",
+                                HttpClient.class.getSimpleName(),
+                                InputStream.class.getSimpleName(),
+                                long.class.getSimpleName()), e);
+                    }
+                } finally {
+                    os.reset();
+                }
+            }
+        };
+
+        ProtocolHandshake.Result result = handshake
+                .createSession(getClient(), command);
+        Dialect dialect = result.getDialect();
+        setCommandCodec(dialect.getCommandCodec());
+        getAdditionalCommands().forEach(this::defineCommand);
+        setResponseCodec(dialect.getResponseCodec());
+        return result.createResponse();
+    }
+
     @Override
-    public Response execute(Command command) throws WebDriverException {
+    public Response execute(Command command) throws WebDriverException, IOException {
```

Comment:
Is it a good idea to make it throwing the checked exception? Id rather wrap IOException into WebDriverException and keep the previous method signature